core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotationUtil.java

```diff
@@ -0,0 +1,358 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.internal.annotation;
+
+import static java.util.Objects.requireNonNull;
+import static org.reflections.ReflectionUtils.getMethods;
+import static org.reflections.ReflectionUtils.withName;
+import static org.reflections.ReflectionUtils.withParametersCount;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Native;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+/**
+ * A utility class which helps to get annotations from an {@link AnnotatedElement}.
+ */
+final class AnnotationUtil {
+    /**
+     * Options to be used for finding annotations from an {@link AnnotatedElement}.
+     */
+    enum FindOption {
+        /**
+         * Get annotations from super classes of the given {@link AnnotatedElement} if the element is a
+         * {@link Class}. Otherwise, this option will be ignored.
+         */
+        LOOKUP_SUPER_CLASSES,
+        /**
+         * Get additional annotations from the meta-annotations which annotate the annotations specified
+         * on the given {@link AnnotatedElement}.
+         */
+        LOOKUP_META_ANNOTATIONS,
+        /**
+         * Collect the annotations specified in the super classes first, the annotations specified on the
+         * given {@link AnnotatedElement} will be collected at last. This option will work only with
+         * the {@link #LOOKUP_SUPER_CLASSES}.
+         */
+        COLLECT_SUPER_CLASSES_FIRST
+    }
+
+    /**
+     * Built-in meta annotations defined in {@code java.lang.annotation} package.
+     */
+    private static final Set<Class<? extends Annotation>> BUILTIN_META_ANNOTATIONS =
+            ImmutableSet.of(Documented.class, Inherited.class, Native.class,
+                            Retention.class, Repeatable.class, Target.class);
+
+    /**
+     * Returns an annotation of the {@code annotationType} or {@link Optional#empty()}, which is found
+     * from the specified {@code element}.
```

Comment:
I find the differences between `find*()`, `find*Declared()`, `find*Inherited()` are not clear looking from the Javadoc. Could you update the method description so that the user of this class understand the differences better?